contracts/transfer/circuits/src/execute.rs

```diff
@@ -0,0 +1,365 @@
+use input::{CircuitInput, WitnessInput};
+use output::{CircuitOutput, WitnessOutput};
+
+use anyhow::{anyhow, Result};
+use dusk_plonk::bls12_381::BlsScalar;
+use dusk_plonk::jubjub::JubJubExtended;
+use phoenix_core::Note;
+use poseidon252::sponge;
+use poseidon252::tree::{self, PoseidonBranch};
+use rand_core::{CryptoRng, RngCore};
+use schnorr::gadgets as schnorr_gadgets;
+
+use dusk_plonk::prelude::*;
+
+mod crossover;
+mod input;
+mod output;
+
+#[cfg(test)]
+mod tests;
+
+/// The circuit responsible for creating a zero-knowledge proof
+/// for a 'send to contract transparent' transaction.
+#[derive(Debug, Default, Clone)]
+pub struct ExecuteCircuit<const DEPTH: usize, const CAPACITY: usize> {
+    pi_positions: Vec<PublicInput>,
+    pub inputs: Vec<CircuitInput<DEPTH>>,
+    pub crossover: CircuitCrossover,
+    pub outputs: Vec<CircuitOutput>,
+    pub tx_hash: BlsScalar,
+}
+
+impl<const DEPTH: usize, const CAPACITY: usize>
+    ExecuteCircuit<DEPTH, CAPACITY>
+{
+    pub const fn transcript_label(&self) -> &'static [u8] {
+        b"execute-circuit"
+    }
+
+    rusk_profile_methods!(self, {
+        format!(
+            "transfer-execute-{}-{}",
+            self.inputs.len(),
+            self.outputs.len()
+        )
+    });
+
+    pub fn set_tx_hash(&mut self, tx_hash: BlsScalar) {
+        self.tx_hash = tx_hash;
+    }
+
+    pub fn add_input<R: RngCore + CryptoRng>(
+        &mut self,
+        rng: &mut R,
+        branch: PoseidonBranch<DEPTH>,
+        sk_r: JubJubScalar,
+        note: Note,
+        value: u64,
+        blinding_factor: JubJubScalar,
+        nullifier: BlsScalar,
+    ) -> Result<()> {
+        let input = CircuitInput::new(
+            rng,
+            branch,
+            sk_r,
+            note,
+            value,
+            blinding_factor,
+            nullifier,
+        );
+
+        self.inputs.push(input);
+
+        Ok(())
+    }
+
+    pub fn set_crossover(
+        &mut self,
+        value_commitment: JubJubExtended,
+        value: u64,
+        blinding_factor: JubJubScalar,
+    ) {
+        self.crossover =
+            CircuitCrossover::new(value_commitment, value, blinding_factor);
+    }
+
+    pub fn add_output(
+        &mut self,
+        note: Note,
+        value: u64,
+        blinding_factor: JubJubScalar,
+    ) {
+        let output = CircuitOutput::new(note, value, blinding_factor);
+        self.outputs.push(output);
+    }
+
+    /// Constant message for the schnorr signature generation
+    ///
+    /// The signature is provided outside the circuit; so that's why it is
+    /// constant
+    ///
+    /// The contents of the message are yet to be defined in the documentation.
+    /// For now, it is treated as a constant.
+    pub const fn sign_message() -> BlsScalar {
+        BlsScalar::one()
+    }
+}
+
+impl<const DEPTH: usize, const CAPACITY: usize> Circuit<'_>
+    for ExecuteCircuit<DEPTH, CAPACITY>
+{
+    circuit_common_methods!(CAPACITY);
```

Comment:
As before, I still believe such things should be a derive macro.

Reply:
Fixed